packages/salesforcedx-vscode-core/src/commands/forceSourceDiff.ts

```diff
@@ -0,0 +1,159 @@
+  ): Promise<void> {
+    const startTime = process.hrtime();
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: getRootWorkspacePath(),
+      env: { SFDX_JSON_TO_STDOUT: 'true' }
+    }).execute(cancellationToken);
+
+    channelService.streamCommandStartStop(execution);
+    channelService.showChannelOutput();
+
+    let stdOut = '';
+    execution.stdoutSubject.subscribe(realData => {
+      stdOut += realData.toString();
+    });
+
+    execution.processExitSubject.subscribe(async () => {
+      this.logMetric(execution.command.logName, startTime);
+
+      try {
+        const diffParser = new DiffResultParser(stdOut);
+        const diffParserSuccess = diffParser.getSuccessResponse();
+        const diffParserError = diffParser.getErrorResponse();
+
+        if (diffParser.isSuccessful() && diffParserSuccess) {
+          const diffResult = diffParserSuccess.result;
+          const remote = vscode.Uri.parse(diffResult.remote);
+          const local = vscode.Uri.parse(diffResult.local);
+          const filename = diffResult.fileName;
+
+          let defaultUsernameorAlias: string | undefined;
+          if (hasRootWorkspace()) {
+            defaultUsernameorAlias = await OrgAuthInfo.getDefaultUsernameOrAlias(
+              false
+            );
+          }
+          vscode.commands.executeCommand(
+            'vscode.diff',
+            remote,
+            local,
+            nls.localize(
+              'force_source_diff_title',
+              defaultUsernameorAlias,
+              filename,
+              filename
+            )
+          );
+        } else if (diffParserError) {
+          channelService.appendLine(diffParserError.message);
+        }
+      } catch (e) {
+        telemetryService.sendException(e.name, e.message);
+        const err = new Error('Error parsing diff result');
+        err.name = 'DiffParserFail';
+        throw err;
+      }
+    });
+
+    notificationService.reportCommandExecutionStatus(
+      execution,
+      cancellationToken
+    );
+    ProgressNotification.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+}
+
+const workspaceChecker = new SfdxWorkspaceChecker();
+
+export class SourcePathGatherer
+  implements ParametersGatherer<{ filePath: string }> {
+  private sourcePath: string;
+
+  public constructor(uri: vscode.Uri) {
+    this.sourcePath = uri.fsPath;
+  }
+
+  public async gather(): Promise<
+    CancelResponse | ContinueResponse<{ filePath: string }>
+  > {
+    return { type: 'CONTINUE', data: { filePath: this.sourcePath } };
+  }
+}
+
+export async function forceSourceDiff(sourceUri: vscode.Uri) {
+  if (sourceUri) {
+    const editor = vscode.window.activeTextEditor;
+    if (
```

Comment:
We'll initially have a limited set of supported files, I'd like to know if there's a better way of limiting/checking for the supported types. I tried updating the `where` clause at the command level but that was getting a bit too complex and there's no easy way to write a test to verify the logic. Any ideas ?

Reply:
Doesn't the CLI command already handle this?

Reply:
Are you referring to when the diff command is exposed in the drop down menu?

Reply:
Could we add this check to the diff plugin itself? We could add a new command for checking if the type is supported. That would save us from being in a situation where the two code bases get out of sync in case we add additional types or remove types.

Reply:
I modified the first implementation of this. In the current version, I'm addressing the scenario where the user has added a keyboard shortcut to the diff command. In that scenario, sourceUri is always null so we need to see if we can get a file uri from the currently open file, if any.

I agree this could get out of sync with the diff plugin but that would be on us to make sure we coordinate the updates/releases instead of leaving a gap where the user cannot use the command in the keyboard shortcut scenario.

Reply:
Is this thread obsolete with these new changes? https://github.com/forcedotcom/salesforcedx-vscode/pull/1568/files/a7a1de7cc9154bc77538ec46ce6b6f3c86b477bb#r315284559

Reply:
Only the first 3 comments, @sfsholden 's comment is still relevant.